lib/statsd/instrument/aggregator.rb

```diff
@@ -0,0 +1,261 @@
+      # @return [void]
+      def increment(name, value = 1, tags: [], no_prefix: false)
+        unless thread_healthcheck
+          sink << datagram_builder(no_prefix: no_prefix).c(name, value, CONST_SAMPLE_RATE, tags)
+          return
+        end
+
+        tags = tags_sorted(tags)
+        key = packet_key(name, tags, no_prefix, COUNT)
+
+        @mutex.synchronize do
+          @aggregation_state[key] ||= 0
+          @aggregation_state[key] += value
+        end
+      end
+
+      def aggregate_timing(name, value, tags: [], no_prefix: false, type: DISTRIBUTION)
+        unless thread_healthcheck
+          sink << datagram_builder(no_prefix: no_prefix).d(name, value, CONST_SAMPLE_RATE, tags)
+          return
+        end
+
+        tags = tags_sorted(tags)
+        key = packet_key(name, tags, no_prefix, type)
+
+        @mutex.synchronize do
+          if @aggregation_state.key?(key) && @aggregation_state[key].size + 1 >= @max_values
+            do_flush
+          end
+          @aggregation_state[key] ||= []
+          @aggregation_state[key] << value
+        end
+      end
+
+      def gauge(name, value, tags: [], no_prefix: false)
+        unless thread_healthcheck
+          sink << datagram_builder(no_prefix: no_prefix).g(name, value, CONST_SAMPLE_RATE, tags)
+          return
+        end
+
+        tags = tags_sorted(tags)
+        key = packet_key(name, tags, no_prefix, GAUGE)
+
+        @mutex.synchronize do
+          @aggregation_state[key] = value
+        end
+      end
+
+      def flush
+        @mutex.synchronize { do_flush }
+      end
+
+      private
+
+      # @return Array<String>
+      EMPTY_ARRAY = [].freeze
+
+      def do_flush
+        @aggregation_state.each do |key, value|
+          case key.type
+          when COUNT
+            @sink << datagram_builder(no_prefix: key.no_prefix).c(
+              key.name,
+              value,
+              CONST_SAMPLE_RATE,
+              key.tags,
+            )
+          when DISTRIBUTION, MEASURE, HISTOGRAM
+            @sink << datagram_builder(no_prefix: key.no_prefix).distribution_value_packed(
+              key.name,
+              key.type.to_s,
+              value,
+              CONST_SAMPLE_RATE,
+              key.tags,
+            )
+          when GAUGE
+            @sink << datagram_builder(no_prefix: key.no_prefix).g(
+              key.name,
+              value,
+              CONST_SAMPLE_RATE,
+              key.tags,
+            )
+          else
+            StatsD.logger.error { "[#{self.class.name}] Unknown aggregation type: #{key.type}" }
+          end
+        end
+        @aggregation_state.clear
+      end
+
+      def tags_sorted(tags)
+        return "" if tags.nil? || tags.empty?
+
+        if tags.is_a?(Hash)
+          tags = tags.sort_by { |k, _v| k.to_s }.map! { |k, v| "#{k}:#{v}" }
+        else
+          tags.sort!
+        end
+        DatagramBuilder.normalize_tags(tags)
```

Comment:
IIRC, calls to `normalize_tags` were inlined elsewhere in the codebase as an optimization - we could keep that knowledge handy in case this shows up as a hot spot in profiles.